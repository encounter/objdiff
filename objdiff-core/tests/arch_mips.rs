use objdiff_core::{diff, obj};

mod common;

#[test]
#[cfg(feature = "mips")]
fn read_mips() {
    let diff_config = diff::DiffObjConfig { mips_register_prefix: true, ..Default::default() };
    let obj =
        obj::read::parse(include_object!("data/mips/main.c.o"), &diff_config, obj::DiffSide::Base)
            .unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "ControlEntry").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "mips")]
fn cross_endian_diff() {
    let diff_config = diff::DiffObjConfig::default();
    let obj_be =
        obj::read::parse(include_object!("data/mips/code_be.o"), &diff_config, obj::DiffSide::Base)
            .unwrap();
    assert_eq!(obj_be.endianness, object::Endianness::Big);
    let obj_le =
        obj::read::parse(include_object!("data/mips/code_le.o"), &diff_config, obj::DiffSide::Base)
            .unwrap();
    assert_eq!(obj_le.endianness, object::Endianness::Little);
    let left_symbol_idx = obj_be.symbols.iter().position(|s| s.name == "func_00000000").unwrap();
    let right_symbol_idx =
        obj_le.symbols.iter().position(|s| s.name == "func_00000000__FPcPc").unwrap();
    let (left_diff, right_diff) =
        diff::code::diff_code(&obj_be, &obj_le, left_symbol_idx, right_symbol_idx, &diff_config)
            .unwrap();
    // Although the objects differ in endianness, the instructions should match.
    assert_eq!(left_diff.instruction_rows[0].kind, diff::InstructionDiffKind::None);
    assert_eq!(right_diff.instruction_rows[0].kind, diff::InstructionDiffKind::None);
    assert_eq!(left_diff.instruction_rows[1].kind, diff::InstructionDiffKind::None);
    assert_eq!(right_diff.instruction_rows[1].kind, diff::InstructionDiffKind::None);
    assert_eq!(left_diff.instruction_rows[2].kind, diff::InstructionDiffKind::None);
    assert_eq!(right_diff.instruction_rows[2].kind, diff::InstructionDiffKind::None);
}

#[test]
#[cfg(feature = "mips")]
fn filter_non_matching() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(
        include_object!("data/mips/vw_main.c.o"),
        &diff_config,
        obj::DiffSide::Base,
    )
    .unwrap();
    insta::assert_debug_snapshot!(obj.symbols);
}
