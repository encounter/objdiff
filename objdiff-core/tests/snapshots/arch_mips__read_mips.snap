---
source: objdiff-core/tests/arch_mips.rs
expression: obj
---
Object {
    arch: ArchMips {
        endianness: Little,
        abi: N32,
        isa_extension: Some(
            R5900EE,
        ),
        ri_gp_value: 16384,
        paired_relocations: [
            {},
            {
                32: 0,
                44: 0,
                48: 0,
                60: 0,
                120: 0,
                124: 0,
                204: 16,
                220: 16,
                296: 8,
                308: 8,
                312: 16,
                316: 16,
                328: 24,
                332: 24,
                344: 32,
                348: 32,
                360: 40,
                364: 40,
                376: 48,
                380: 48,
                392: 56,
                396: 56,
                408: 0,
                412: 0,
            },
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
        ],
        ignored_symbols: {},
    },
    endianness: Little,
    symbols: [
        Symbol {
            name: "[.text]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                0,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.data]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.bss]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                3,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.mdebug.eabi64]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                6,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.sdata]",
            demangled_name: None,
            address: 0,
            size: 64,
            kind: Section,
            section: Some(
                8,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.rodata]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                7,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.reginfo]",
            demangled_name: None,
            address: 0,
            size: 24,
            kind: Section,
            section: Some(
                4,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.mdebug]",
            demangled_name: None,
            address: 0,
            size: 1932,
            kind: Section,
            section: Some(
                5,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "gcc2_compiled.",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: Some(
                0,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "__gnu_compiled_c",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: Some(
                0,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "PacketBottomNewVu1DropMicroCode",
            demangled_name: None,
            address: 0,
            size: 12,
            kind: Object,
            section: Some(
                7,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "ControlEntry",
            demangled_name: None,
            address: 0,
            size: 184,
            kind: Function,
            section: Some(
                0,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglSleep",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "WorkEnd",
            demangled_name: None,
            address: 64,
            size: 4,
            kind: Object,
            section: Some(
                8,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglSoundLoadEffect",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglSoundLoadSwd",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "SsdAddWaveData",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "SsdSpuDmaCompleted",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglRenderDispOn",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglCdLoadOverlay",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "LogoFirst",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "Title",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "InitializeSystem",
            demangled_name: None,
            address: 184,
            size: 356,
            kind: Function,
            section: Some(
                0,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceSifInitRpc",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceCdInit",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceSifRebootIop",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceSifSyncIop",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceSifInitIopHeap",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceSifLoadFileReset",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceCdMmode",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceFsReset",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglCdSifLoadModule",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglCdPowerOffCB",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "sceCdPOffCallback",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglSoundInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglCdInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglPadInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglMcInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglTaskInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglDmaInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglGeometryInit",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglPacketInit",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglRenderInit",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglFontInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglMovieInit",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglMenuInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "main",
            demangled_name: None,
            address: 544,
            size: 88,
            kind: Function,
            section: Some(
                0,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "__main",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "main_param_argc",
            demangled_name: None,
            address: 68,
            size: 4,
            kind: Object,
            section: Some(
                8,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "BootDisplay",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "main_param_argv",
            demangled_name: None,
            address: 72,
            size: 4,
            kind: Object,
            section: Some(
                8,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglThreadInitial",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "xglThreadRotate",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
    ],
    sections: [
        Section {
            id: ".text-0",
            name: ".text",
            address: 0,
            size: 632,
            kind: Code,
            data: SectionData(
                632,
            ),
            flags: FlagSet(),
            align: Some(
                8,
            ),
            relocations: [
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 20,
                    target_symbol: 12,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        7,
                    ),
                    address: 28,
                    target_symbol: 13,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 32,
                    target_symbol: 4,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 40,
                    target_symbol: 14,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 44,
                    target_symbol: 4,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 48,
                    target_symbol: 10,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        7,
                    ),
                    address: 52,
                    target_symbol: 13,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 56,
                    target_symbol: 15,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 60,
                    target_symbol: 10,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        7,
                    ),
                    address: 64,
                    target_symbol: 13,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 72,
                    target_symbol: 16,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 80,
                    target_symbol: 17,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 96,
                    target_symbol: 18,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 104,
                    target_symbol: 19,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 112,
                    target_symbol: 20,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 120,
                    target_symbol: 21,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 124,
                    target_symbol: 21,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 128,
                    target_symbol: 19,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 136,
                    target_symbol: 21,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 160,
                    target_symbol: 19,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 200,
                    target_symbol: 23,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 204,
                    target_symbol: 5,
                    addend: 16,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 208,
                    target_symbol: 24,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 216,
                    target_symbol: 25,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 220,
                    target_symbol: 5,
                    addend: 16,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 232,
                    target_symbol: 26,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 248,
                    target_symbol: 23,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 256,
                    target_symbol: 27,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 264,
                    target_symbol: 28,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 272,
                    target_symbol: 24,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 280,
                    target_symbol: 29,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 288,
                    target_symbol: 30,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 296,
                    target_symbol: 4,
                    addend: 8,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 304,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 308,
                    target_symbol: 4,
                    addend: 8,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 312,
                    target_symbol: 4,
                    addend: 16,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 316,
                    target_symbol: 4,
                    addend: 16,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 320,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 328,
                    target_symbol: 4,
                    addend: 24,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 332,
                    target_symbol: 4,
                    addend: 24,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 336,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 344,
                    target_symbol: 4,
                    addend: 32,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 348,
                    target_symbol: 4,
                    addend: 32,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 352,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 360,
                    target_symbol: 4,
                    addend: 40,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 364,
                    target_symbol: 4,
                    addend: 40,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 368,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 376,
                    target_symbol: 4,
                    addend: 48,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 380,
                    target_symbol: 4,
                    addend: 48,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 384,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 392,
                    target_symbol: 4,
                    addend: 56,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 396,
                    target_symbol: 4,
                    addend: 56,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 400,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        5,
                    ),
                    address: 408,
                    target_symbol: 32,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 412,
                    target_symbol: 32,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 416,
                    target_symbol: 33,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 428,
                    target_symbol: 34,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        7,
                    ),
                    address: 432,
                    target_symbol: 13,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 436,
                    target_symbol: 35,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 444,
                    target_symbol: 36,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 452,
                    target_symbol: 37,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 468,
                    target_symbol: 38,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 476,
                    target_symbol: 39,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 484,
                    target_symbol: 40,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 492,
                    target_symbol: 41,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 500,
                    target_symbol: 42,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 508,
                    target_symbol: 43,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 516,
                    target_symbol: 44,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 532,
                    target_symbol: 45,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 564,
                    target_symbol: 47,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        7,
                    ),
                    address: 572,
                    target_symbol: 48,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 576,
                    target_symbol: 49,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        7,
                    ),
                    address: 580,
                    target_symbol: 50,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 584,
                    target_symbol: 22,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 592,
                    target_symbol: 51,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 600,
                    target_symbol: 52,
                    addend: 0,
                },
            ],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".rel.text-0",
            name: ".rel.text",
            address: 0,
            size: 608,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                4,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".data-0",
            name: ".data",
            address: 0,
            size: 0,
            kind: Data,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".bss-0",
            name: ".bss",
            address: 0,
            size: 0,
            kind: Bss,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".reginfo-0",
            name: ".reginfo",
            address: 0,
            size: 24,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                4,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".mdebug-0",
            name: ".mdebug",
            address: 0,
            size: 1932,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                4,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".mdebug.eabi64-0",
            name: ".mdebug.eabi64",
            address: 0,
            size: 0,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".rodata-0",
            name: ".rodata",
            address: 0,
            size: 43,
            kind: Data,
            data: SectionData(
                43,
            ),
            flags: FlagSet(),
            align: Some(
                8,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".sdata-0",
            name: ".sdata",
            address: 0,
            size: 76,
            kind: Data,
            data: SectionData(
                76,
            ),
            flags: FlagSet(),
            align: Some(
                8,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".symtab-0",
            name: ".symtab",
            address: 0,
            size: 864,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                4,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".strtab-0",
            name: ".strtab",
            address: 0,
            size: 653,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".shstrtab-0",
            name: ".shstrtab",
            address: 0,
            size: 95,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
    ],
    split_meta: None,
    path: None,
    timestamp: None,
}
